### PR TITLE
Fix platform-dependent FS stats on Darwin

### PR DESCRIPTION
**Pull request by julio-lopez**

### CODE REVIEW

**Comment by julio-lopez** [comment]
Ping

**Comment by julio-lopez** [comment]
Folks, any feedback? What's the blocker? Please let me know if there is anything that should be changed or improved. Thanks.

File: internal/fsstat/fsstat_linux.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+				Nanos:   int32(stat.Mtim.Nsec),
+			},
+			Ctime: &tspb.Timestamp{
+				Seconds: stat.Ctim.Sec,
+				Nanos:   int32(stat.Ctim.Nsec),
+			},
+		}
+	}
+
+	log.Panicf("unexpected info.Sys() type %T for %#v", info.Sys(), info)
```
Comment (x1ddos) [inline internal/fsstat/fsstat_linux.go]: I'd prefer to return an error instead of panic.

File: internal/fsstat/fsstat_linux.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package fsstat
+
+import (
+	"log"
+	"os"
+	"syscall"
+
+	tspb "github.com/golang/protobuf/ptypes/timestamp"
+
```
Comment (x1ddos) [inline internal/fsstat/fsstat_linux.go]: no need for empty line here. it's ok to combine proto imports together.

File: internal/fsstat/fsstat_darwin.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+			Size:    stat.Size,
+			Blksize: int64(stat.Blksize),
+			Blocks:  stat.Blocks,
+			Atime:   timespec2Timestamp(stat.Atimespec),
+			Mtime:   timespec2Timestamp(stat.Mtimespec),
+			Ctime:   timespec2Timestamp(stat.Ctimespec),
+		}
+	}
+
+	log.Panicf("unexpected info.Sys() type %T for %#v", info.Sys(), info)
```
Comment (x1ddos) [inline internal/fsstat/fsstat_darwin.go]: Also here, I'd return an error instead.
Worst case, the caller could panic so at least it's in one place.

File: internal/fsstat/dev.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package fsstat
```
Comment (x1ddos) [inline internal/fsstat/dev.go]: A package doc comment would be nice, even a short one.

File: internal/fsstat/fsstat_darwin.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package fsstat
+
+import (
+	"log"
+	"os"
+	"syscall"
+
+	tspb "github.com/golang/protobuf/ptypes/timestamp"
+
```
Comment (x1ddos) [inline internal/fsstat/fsstat_darwin.go]: No need for empty line.

File: internal/fsstat/fsstat_linux.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+			Inode:   stat.Ino,
+			Nlink:   stat.Nlink,
+			Mode:    stat.Mode,
+			Uid:     stat.Uid,
+			Gid:     stat.Gid,
+			Rdev:    stat.Rdev,
+			Size:    stat.Size,
+			Blksize: stat.Blksize,
+			Blocks:  stat.Blocks,
+			Atime: &tspb.Timestamp{
```
Comment (x1ddos) [inline internal/fsstat/fsstat_linux.go]: Would be nice to reuse `timespec2Timestamp(...)` here in the other places below.

File: internal/fsstat/fsstat_darwin.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	fspb "github.com/google/fswalker/proto/fswalker"
+)
+
+// ToStat returns a fspb.ToStat with the file info from the given file
+func ToStat(info os.FileInfo) *fspb.FileStat {
+	if stat, ok := info.Sys().(*syscall.Stat_t); ok {
+		return &fspb.FileStat{
+			Dev:     uint64(stat.Dev),
```
Comment (x1ddos) [inline internal/fsstat/fsstat_darwin.go]: What's the deal with `uint64(stat.Dev)` here in darwin version but no such thing in linux, just `Dev: stat.Dev`?

File: walker.go [hunk 8]
```diff
@@ -250,11 +223,10 @@ func (w *Walker) worker(ctx context.Context, chPaths <-chan string) error {
 		if err != nil {
 			return fmt.Errorf("unable to get file info for base path %q: %v", path, err)
 		}
-		baseStat, ok := baseInfo.Sys().(*syscall.Stat_t)
-		if !ok {
-			return fmt.Errorf("unable to get file stat on base path: %q", path)
+		baseDev, err := fsstat.DevNumber(baseInfo)
+		if err != nil {
+			return fmt.Errorf("unable to get file stat on base path: %q, %v", path, err)
```
Comment (x1ddos) [inline walker.go]: "... on base path %q: %v"

File: internal/fsstat/fsstat_darwin.go (hunk 7 above)
Comment (julio-lopez) [reply internal/fsstat/fsstat_darwin.go]: Different sizes for the definitions for Darwin and Linux:

Darwin:
```go
type Stat_t struct {
	Dev           int32
	Mode          uint16
	Nlink         uint16
	Ino           uint64
	Uid           uint32
	Gid           uint32
	Rdev          int32
	Pad_cgo_0     [4]byte
	Atimespec     Timespec
	Mtimespec     Timespec
	Ctimespec     Timespec
	Birthtimespec Timespec
	Size          int64
	Blocks        int64
	Blksize       int32
	Flags         uint32
	Gen           uint32
	Lspare        int32
	Qspare        [2]int64
}
```

Linux:
```go
type Stat_t struct {
	Dev       uint64
	Ino       uint64
	Nlink     uint64
	Mode      uint32
	Uid       uint32
	Gid       uint32
	X__pad0   int32
	Rdev      uint64
	Size      int64
	Blksize   int64
	Blocks    int64
	Atim      Timespec
	Mtim      Timespec
	Ctim      Timespec
	X__unused [3]int64
}
```

Comment (x1ddos) [reply internal/fsstat/fsstat_darwin.go]: Ah of course. Makes sense.

File: walker_test.go [hunk 9]
```diff
@@ -290,12 +290,20 @@ func TestConvert(t *testing.T) {
 		},
 	}
 
-	gotFile := wlkr.convert(path, nil) // ensuring there is no problems with nil file stats.
+	gotFile, err := wlkr.convert(path, nil) // ensuring there is no problems with nil file stats.
+	if err != nil {
+		t.Fatal("unexpected error:", err)
```
Comment (x1ddos) [inline walker_test.go]: Just `t.Fatal(err)`. The fact that it fails a test is already unexpected.

File: walker_test.go [hunk 10]
```diff
@@ -290,12 +290,20 @@ func TestConvert(t *testing.T) {
...
+	}
+
 	if wantFile.Path != gotFile.Path {
 		t.Errorf("convert() path = %q; want: %q", gotFile.Path, wantFile.Path)
 	}
 
-	gotFile = wlkr.convert(path, info)
+	gotFile, err = wlkr.convert(path, info)
+	if err != nil {
+		t.Fatal("unexpected error:", err)
```
Comment (x1ddos) [inline walker_test.go]: Also here, simply `t.Fatal(err)` or add context like `t.Fatalf("wlkr.convert: %v", err)`. A failing test is already unexpected on its own.

File: internal/fsstat/dev.go [hunk 11]
```diff
@@ -0,0 +1,23 @@
+// Package fsstat provides access to platform specific file stat info
```
Comment (x1ddos) [inline internal/fsstat/dev.go]: Trailing dot `.` to complete the sentence, please.
Here and in other doc comments.
https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

Comment (julio-lopez) [reply internal/fsstat/dev.go]:
```suggestion
// Package fsstat provides access to platform specific file stat info.
```

File: walker_test.go (hunk 9 above)
Comment (julio-lopez) [reply walker_test.go]:
```suggestion
		t.Fatal(err)
```

File: walker_test.go (hunk 10 above)
Comment (julio-lopez) [reply walker_test.go]:
```suggestion
		t.Fatal(err)
```

**Comment by julio-lopez** [comment]
Any additional changes required?

**Pull request merged**